### PR TITLE
make src into a range and get rid of count

### PR DESCRIPTION
**Pull request by oconnor663**
This is an example implementation of a change @SimonSapin suggested in https://github.com/rust-lang/rust/pull/53652#issuecomment-419355935.

### CODE REVIEW

**Comment by SimonSapin** [comment]
(Note that this is ~half of my suggestion, I also proposed changing `dest` from `usize` to `RangeFrom<usize>` to make it clear that it is the start position of some range. (However it is *not* `dest.start()..buffer.len()` which is the usual meaning of `RangeFrom`, so maybe that’s not such a good idea…)

**Comment by oconnor663** [comment]
Here's an example instantiation to look at the difference: https://godbolt.org/z/PsxH7K

It seems like a few instructions get added to the caller in this case, probably too small of a difference to be measurable?

File: src/lib.rs [hunk 1]
```diff
@@ -46,17 +48,27 @@
...
-    assert!(src <= slice.len() - count);
-    assert!(dest <= slice.len() - count);
+pub fn copy_in_place<T: Copy, R: RangeBounds<usize>>(slice: &mut [T], src: R, dest: usize) {
+    let src_start = match src.start_bound() {
+        Bound::Included(&n) => n,
+        Bound::Excluded(&n) => n.saturating_add(1),
+        Bound::Unbounded => 0,
+    };
+    let src_end = match src.end_bound() {
+        Bound::Included(&n) => n.saturating_add(1),
```
Comment (oconnor663) [inline src/lib.rs]: I wonder if this is technically correct. If a slice actually ends at the largest possible address (I think that's one less than the maximum representable, to allow C to compute the `p + len` pointer?), then an inclusive range that passes the end of the slice by one would be converted from the panic it should be to a valid range one element shorter than the caller asked for.

**Comment by oconnor663** [comment]
Playing with this a little more, I think it's obviously better to read and write than my original API. I'm going to land it here and update the Rust PR.

**Pull request merged**